### PR TITLE
Make substring null behavior congruent with tsql

### PR DESCRIPTION
**Pull request by jusjosj**
### Description

This commit addresses a difference in the way that substring handled null arguments between babelfish and tsql.  Prior to this change, babelfish would raise an error on any null input to the substring function regardless of whether or not this is a null constant or another data type with the value of null.  After this change, babelfish will only raise an error on values which are considered to be null constants restoring paraity to tsql
 
Task:BABEL-3347
Signed-off-by: Justin Jossick [jusjosj@amazon.com](mailto:jusjosj@amazon.com)

### Issues Resolved

[BABEL-3347] 

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/string.c [hunk 1]
```diff
@@ -478,17 +478,32 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
...
+	FuncExpr *fe;
+	Node *node;
+
 	if (PG_ARGISNULL(0))
 	{
-		/* Raise error if first arg is null */
-		ereport(ERROR,
-				(errcode(ERRCODE_SUBSTRING_ERROR),
-					errmsg("Argument data type NULL is invalid for argument 1 of substring function")));
+		fe = (FuncExpr *)fcinfo->flinfo->fn_expr;
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/string.c]: To be safer, let's add some null/NIL checks, e.g., `fe` and `fe->args`.

File: contrib/babelfishpg_tsql/src/string.c [hunk 2]
```diff
@@ -478,17 +478,32 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
...
+		node = (Node *)fe->args->elements[0].ptr_value;
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/string.c]: Can we use a wrapper function in pg_list.h, e.g., `linitial_node`?

File: contrib/babelfishpg_tsql/src/string.c [hunk 3]
```diff
@@ -478,17 +478,32 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
...
+		/* Handle null literal */
+		if (node->type == T_Const) {
+			ereport(ERROR,
+					(errcode(ERRCODE_SUBSTRING_ERROR),
+						errmsg("Argument data type NULL is invalid for argument 1 of substring function")));
+		}
+		/* Handle other cases for null such as table variables */
+		PG_RETURN_NULL();
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/string.c]: Could you also add a test for `cast(null as varchar)` case?

File: contrib/babelfishpg_tsql/src/string.c [hunk 4]
```diff
@@ -478,17 +478,32 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
...
 	}
+
 	if (PG_ARGISNULL(1) || PG_ARGISNULL(2))
 		PG_RETURN_NULL();
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/string.c:486]: Please add a test where the 2nd or the 3rd argument is null from table variables.

File: test/JDBC/expected/BABEL-3347-vu-prepare.out [hunk 5]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE babel_3347_table(text_test VARCHAR(100));
```
Comment (2jungkook) [inline test/JDBC/expected/BABEL-3347-vu-prepare.out]: If possible, let's add this test to other version's schedule files.

File: test/JDBC/expected/BABEL-3347-vu-verify.out [hunk 6]
```diff
@@ -0,0 +1,26 @@
+-- table variable
+SELECT SUBSTRING(text_test, 2,3) FROM babel_3347_table
```
Comment (2jungkook) [inline test/JDBC/expected/BABEL-3347-vu-verify.out]: The one of the main purposes of version upgrade test is to make sure all objects are correctly migrated. Please create a view for each `SELECT` statement.

File: contrib/babelfishpg_tsql/src/string.c (hunk 4 above)
Comment (jusjosj) [reply contrib/babelfishpg_tsql/src/string.c:486]: I can, but this is already covered in the tests for babel-1114 so I'm not sure that it is necessary

File: contrib/babelfishpg_tsql/src/string.c [hunk 7]
```diff
@@ -478,17 +478,40 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
...
+
 	if (PG_ARGISNULL(0))
 	{
-		/* Raise error if first arg is null */
-		ereport(ERROR,
-				(errcode(ERRCODE_SUBSTRING_ERROR),
-					errmsg("Argument data type NULL is invalid for argument 1 of substring function")));
+		fe = (FuncExpr *)fcinfo->flinfo->fn_expr;
+
+		if (fe == NULL || fe->args == NULL) {
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/string.c]: For a list, we use `NIL` instead of `NULL`.

File: contrib/babelfishpg_tsql/src/string.c [hunk 8]
```diff
@@ -478,17 +478,40 @@ prepare_format_string(StringInfo buf, char *msg_string, int nargs,
...
+			ereport(ERROR,
+					(errcode(ERRCODE_SUBSTRING_ERROR),
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/string.c]: Let's use `ERRCODE_INTERNAL_ERROR`.

File: test/JDBC/expected/BABEL-3347-vu-cleanup.out [hunk 9]
```diff
@@ -0,0 +1,26 @@
+DROP VIEW IF EXISTS babel_3347_vu1
```
Comment (2jungkook) [inline test/JDBC/expected/BABEL-3347-vu-cleanup.out:1]: Here, `vu` means `version_upgrade`. So the naming convention would be `babel_3347_vu_v1`.

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 10]
```diff
@@ -2153,6 +2153,29 @@ class tsqlMutator : public TSqlParserBaseListener
...
+	{
+		if (ctx->func_proc_name_server_database_schema())
+		{
+			auto fpnsds = ctx->func_proc_name_server_database_schema();
+
+			if (fpnsds->DOT().empty() && fpnsds->id().back()->keyword()) /* built-in functions */
+			{
+				auto id = fpnsds->id().back();
+
+				if (id->keyword()->SUBSTRING()) /* NULLIF */
```
Comment (2jungkook) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: 1. Please fix the comment
2. We need to fix NULLIF (and probably for CHECKSUM) as well. Please create a JIRA for that.

**Review by 2jungkook — APPROVED** [review]

**Pull request merged**